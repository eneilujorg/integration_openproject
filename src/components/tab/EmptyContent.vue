<template>
	<div class="empty-content">
		<div class="empty-content--wrapper">
			<div class="empty-content--icon">
				<img :src="flowSvg" alt="flow">
			</div>
			<div v-if="adminConfigStatus" class="empty-content--title">
				{{ emptyContentMessage }}
			</div>
			<div v-if="showConnectButton" class="empty-content--connect-button">
				<OAuthConnectButton :request-url="requestUrl" :admin-config-status="adminConfigStatus" />
			</div>
		</div>
	</div>
</template>

<script>
import { generateUrl } from '@nextcloud/router'
import { translate as t } from '@nextcloud/l10n'
import OAuthConnectButton from '../OAuthConnectButton'

export default {
	name: 'EmptyContent',
	components: { OAuthConnectButton },
	props: {
		state: {
			type: String,
			required: true,
			default: 'ok',
		},
		requestUrl: {
			type: String,
			required: true,
		},
		errorMessage: {
			type: String,
			default: '',
		},
		adminConfigStatus: {
			type: Boolean,
			required: true,
		},
	},
	data() {
		return {
			settingsUrl: generateUrl('/settings/user/connected-accounts'),
		}
	},
	computed: {
		flowSvg() {
			return require('../../../img/flow.svg')
		},
		showConnectButton() {
			return ['error', 'no-token'].includes(this.state)
		},
		emptyContentMessage() {
			if (this.state === 'no-token') {
				return t('integration_openproject', 'No OpenProject account connected')
			} else if (this.state === 'connection-error') {
				return t('integration_openproject', 'Error connecting to OpenProject')
			} else if (this.state === 'failed-fetching-workpackages') {
				return t('integration_openproject', 'Could not fetch work packages from OpenProject')
			} else if (this.state === 'error') {
				return t('integration_openproject', 'Unexpected Error')
			} else if (this.state === 'ok') {
				return t('integration_openproject', 'No workspaces linked yet, search for work package to add!')
			}
			return 'invalid state'
		},
	},
}
</script>

<style scoped lang="scss">
.empty-content {
	height: 100%;
	display: flex;
	align-items: center;
	justify-content: center;
	&--wrapper {
		height: fit-content;
	}
	&--icon {
		padding: 1vh 0;
		display: flex;
		align-items: center;
		justify-content: center;
		img {
			height: 50px;
			width: 50px;
		}
	}
	&--title {
		text-align: center;
		font-size: 1.2rem;
		line-height: 1.4rem;
		font-weight: 600;
		padding-top: 4px;
		color: #333333;
	}
	&--subtitle {
		font-size: .875rem;
		font-weight: 400;
		color: #6d6d6d;
		line-height: 1rem;
		padding: 8px 0;
	}
	&--connect-button {
		padding: 1vh 0;
		display: flex;
		align-items: center;
		justify-content: center;
	}
}
</style>
