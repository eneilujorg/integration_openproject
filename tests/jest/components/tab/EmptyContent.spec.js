/* jshint esversion: 8 */
import { shallowMount, createLocalVue } from '@vue/test-utils'
import EmptyContent from '../../../../src/components/tab/EmptyContent'
const localVue = createLocalVue()

describe('EmptyContent.vue Test', () => {
	let wrapper
	const emptyContentMessageSelector = '.empty-content--message'
	const connectButtonSelector = 'oauthconnectbutton-stub'

	describe('connect button', () => {
		it.each([{
			state: 'ok',
			viewed: false,
		}, {
			state: 'no-token',
			viewed: true,
		}, {
			state: 'error',
			viewed: true,
		}])('should be displayed depending on the state', (cases) => {
			wrapper = getWrapper({ state: cases.state })
			expect(wrapper.find(connectButtonSelector).exists()).toBe(cases.viewed)
		})
	})
	describe('content title', () => {
		it('should not be displayed if the request url is not valid', () => {
			wrapper = getWrapper({ requestUrl: false })
			expect(wrapper.find(emptyContentMessageSelector).exists()).toBe(false)
		})
		it.each([{
			state: 'no-token',
		}, {
			state: 'error',
		}, {
			state: 'connection-error',
		}, {
			state: 'failed-fetching-workpackages',
		}, {
			state: 'ok',
		}])('shows the correct empty message depending on states if the request url is valid', async (cases) => {
			wrapper = getWrapper({ state: cases.state, adminConfigStatus: true })
			expect(wrapper.find(emptyContentMessageSelector).exists()).toBeTruthy()
			expect(wrapper.find(emptyContentMessageSelector)).toMatchSnapshot()
		})
	})
})

function getWrapper(propsData = {}) {
	return shallowMount(EmptyContent, {
		localVue,
		mocks: {
			t: (msg) => msg,
		},
		propsData: {
			state: 'ok',
			requestUrl: 'http://openproject/',
			...propsData,
		},
	})
}
