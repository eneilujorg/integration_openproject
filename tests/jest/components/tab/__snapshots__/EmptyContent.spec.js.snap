// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmptyContent.vue Test content title shows the correct empty message depending on states if the request url is valid 1`] = `
<div class="empty-content--message">
  <div class="empty-content--message--title">
    No connection with OpenProject
  </div>
  <!---->
</div>
`;

exports[`EmptyContent.vue Test content title shows the correct empty message depending on states if the request url is valid 2`] = `
<div class="empty-content--message">
  <div class="empty-content--message--title">
    Unexpected Error
  </div>
  <!---->
</div>
`;

exports[`EmptyContent.vue Test content title shows the correct empty message depending on states if the request url is valid 3`] = `
<div class="empty-content--message">
  <div class="empty-content--message--title">
    Error connecting to OpenProject
  </div>
  <!---->
</div>
`;

exports[`EmptyContent.vue Test content title shows the correct empty message depending on states if the request url is valid 4`] = `
<div class="empty-content--message">
  <div class="empty-content--message--title">
    Could not fetch work packages from OpenProject
  </div>
  <!---->
</div>
`;

exports[`EmptyContent.vue Test content title shows the correct empty message depending on states if the request url is valid 5`] = `
<div class="empty-content--message">
  <div class="empty-content--message--title">
    No OpenProject links yet
  </div>
  <div class="empty-content--message--sub-title">
    To add a link, use the search bar above to find the desired work package
  </div>
</div>
`;
